File: src/world/mood.ts
// World層に渡す「気分」の型と、両層の合意事項（docs/UI-SPEC.md §6.6）。
//
// scenes.ts から切り出してある。ここは three.js に依存しないので、
// **画面もテストも import できる**——属性の並びは 3D の見た目を決める
// 合意事項なので、ブラウザを立ち上げずに固定できることに意味がある。

/**
 * 画面ごとの「気分」（改善指示書 §3）。
 *
 * 3D 側は画面の**状態**にも反応する。派遣先を選び直せば入口の色が変わるし、
 * 誰かが潜っていれば拠点の灯りが変わる。文字は一切持たないまま、
 * 「今どうなっているか」を光と密度で言う。
 *
 * 画面が three.js に触らずに済むよう、渡すのは**数値だけ**にする。
 */
export interface Mood {
  /** 主となる色（派遣先の属性・レアリティなど） */
  accent?: number;
  /** 0〜1。深さ・強さ。塵の密度や光の強さに効く */
  intensity?: number;
  /** 0〜1。人の気配（拠点の灯り）。1 なら全員在宅 */
  presence?: number;
  /** 畑の中身。要素数がそのまま「開いている枠の数」になる */
  slots?: readonly PlotMood[];
  /** 次の枠を買えるか。買えるときだけ温室に「＋」が立つ */
  canExpand?: boolean;
}

/**
 * 畑1枠の状態。
 *
 * **ここも数値だけで喋る。** 薬草の ID を渡せば楽だが、そうすると
 * World層がゲームの語彙（`'ironleaf'`）を知ることになり、
 * 「光と形しか持たない層」という約束が崩れる。属性の**番号**だけ渡す。
 */
export interface PlotMood {
  /** 属性の番号（MOOD_ELEMENTS の添字）。-1 なら空き */
  kind: number;
  /** 育ち具合 0〜1 */
  ratio: number;
}

/**
 * Mood が運ぶ属性の並び。**両層の唯一の合意事項**。
 *
 * 添字で受け渡すので、並びが狂うと「火苔を植えたのに青い蕾が生える」
 * という壊れ方をする。しかも画面は普通に動いてしまうので気づけない。
 * 並び順に意味があることを試験で固定しておく
 * （test/run-tests.ts「薬草の見た目」）。
 */
export const MOOD_ELEMENTS = ['physical', 'fire', 'ice', 'lightning', 'poison'] as const;
export type MoodElement = typeof MOOD_ELEMENTS[number];

/** 属性名 → 番号。知らない名前は物理に寄せる（3Dが消えるよりは鈍く出るほうがよい） */
export function elementIndex(e: string): number {
  const i = MOOD_ELEMENTS.indexOf(e as MoodElement);
  return i < 0 ? 0 : i;
}

File: ios/Delvers/World/Mood.swift
import DelversCore
import SwiftUI

// 画面から 3D へ渡すもの（web 版 docs/UI-SPEC.md §6.6 の iOS 版）。
//
// **渡すのは数値だけ。** 文字は渡さない——World層は光と密度でしか喋らない。
// ここに文字列を1つ通した瞬間、3D 側が「何の画面か」を知ることになり、
// 2層に分けた意味が消える。

struct PlotMood: Equatable {
    /// 属性の番号（-1 は空き）
    var kind: Int
    /// 育ち具合 0〜1
    var ratio: Double
}

struct NodeMood: Equatable {
    /// 0＝未解放 / 1＝解放済み / 2＝踏破済み
    var state: Int
    /// 属性の番号。複合は -1
    var element: Int
}

struct PropMood: Equatable {
    var chest: Double = 0
    var mail: Double = 0
    var sign: Double = 0
    var shelf: Double = 0
}

struct Mood: Equatable {
    /// 主となる色
    var accent: Color = DS.gold
    /// 0〜1。深さ・強さ
    var intensity: Double = 0.4
    /// 0〜1。人の気配（拠点の灯り）
    var presence: Double = 1
    /// 畑の中身。要素数がそのまま開いている枠の数
    var slots: [PlotMood] = []
    /// 畑を広げられるか
    var canExpand = false
    /// 地図のノード
    var nodes: [NodeMood] = []
    /// 選んでいるノード
    var selected: Int = -1
    /// 拠点に置いた物の状態
    var props = PropMood()
}

/// 属性 → 番号。両層の唯一の合意事項（DelversCore の並びに合わせる）
func moodElementIndex(_ e: Element) -> Int {
    switch e {
    case .physical: return 0
    case .fire: return 1
    case .lightning: return 2
    case .poison: return 3
    case .ice: return 4
    }
}

/// 番号 → 光の色。World 層が持つ唯一の対応表
func moodElementColor(_ i: Int) -> UIColor {
    switch i {
    case 1: return UIColor(red: 1.0, green: 0.51, blue: 0.28, alpha: 1)   // 炎
    case 2: return UIColor(red: 0.91, green: 0.75, blue: 0.45, alpha: 1)  // 雷
    case 3: return UIColor(red: 0.49, green: 0.86, blue: 0.54, alpha: 1)  // 毒
    case 4: return UIColor(red: 0.44, green: 0.78, blue: 1.0, alpha: 1)   // 氷
    default: return UIColor(red: 0.62, green: 0.69, blue: 0.82, alpha: 1) // 物理
    }
}
